apps/vscode/src/providers/context-keys.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * context-keys.ts
+ *
+ * Copyright (C) 2024 by Posit Software, PBC
+ *
+ * Unless you have received this program directly from Posit Software pursuant
+ * to the terms of a commercial license agreement with Posit Software, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+
+import * as vscode from "vscode";
+import debounce from "lodash.debounce";
+
+import { isQuartoDoc } from "../core/doc";
+import { MarkdownEngine } from "../markdown/engine";
+import { mainLanguage } from "../vdoc/vdoc";
+
+const debounceOnDidChangeDocumentMs = 250;
+
+export function activateContextKeySetter(
+  context: vscode.ExtensionContext,
+  engine: MarkdownEngine
+) {
+
+  // set context keys when docs are opened
+  vscode.workspace.onDidOpenTextDocument(
+    (doc) => {
+      if (doc === vscode.window.activeTextEditor?.document) {
+        if (isQuartoDoc(doc)) {
+          setContextKeys(vscode.window.activeTextEditor, engine);
+        }
+      }
+    },
+    null,
+    context.subscriptions
+  );
+
+  // set context keys when visible text editors change
+  vscode.window.onDidChangeVisibleTextEditors(
+    (_editors) => {
+      triggerUpdateContextKeys(engine);
+    },
+    null,
+    context.subscriptions
+  );
+
+  // set context keys on changes to the document (if its visible)
+  vscode.workspace.onDidChangeTextDocument(
+    (event) => {
+      const visibleEditor = vscode.window.visibleTextEditors.find(editor => {
+        return editor.document.uri.toString() === event.document.uri.toString();
+      });
+      if (visibleEditor) {
+        debounce(
+          () => setContextKeys(visibleEditor, engine),
+          debounceOnDidChangeDocumentMs
+        )();
+      }
+    },
+    null,
+    context.subscriptions
+  );
+
+  // set context keys at activation time
+  triggerUpdateContextKeys(engine);
+
+}
+
+function triggerUpdateContextKeys(engine: MarkdownEngine) {
+  for (const editor of vscode.window.visibleTextEditors) {
+    setContextKeys(editor, engine);
+  }
+}
```

Comment:
I think this is looking great!

My main comment is about this, and any other usage of `visibleTextEditors`

IIUC, the context key is "global". i.e. there is _not_ a context key per document, there is exactly 1 context key set globally.

If that is the case, then the context key should probably only be set for the _active_ text editor, not for _all visible_ text editors (because only 1 wins!).

For example, if I open 2 qmds side by side in split panes, one that is mostly R and one that is mostly Python, then I'm not entirely sure which is going to win here because both are "visible". However, only one at a time is ever "active", and I feel like that is probably what we should be targeting with this context key.

Is it possible all we really need to do is having a single hook for `onDidChangeActiveTextEditor()`?

Does that make any sense?

Reply:
Makes _total_ sense, yep. Changed in 8d00a51d44519ed9d4e47abdc47122e3c8e07151.

Reply:
I added the hook for `onDidChangeTextDocument()` for the case where someone changes their document from being Python to R or vice versa. That is certainly rare, but it feels fairly broken if we don't include it.

Reply:
You are correct that we don't need `onDidOpenTextDocument()` if we have `onDidChangeActiveTextEditor()`. 👍

Reply:
> I added the hook for onDidChangeTextDocument()

Makes sense because you also dont have a "main" language at the start of a document until you actually start typing and adding some chunks (or maybe there is a default, but it likely isnt right)

---

PR looks great nice job!